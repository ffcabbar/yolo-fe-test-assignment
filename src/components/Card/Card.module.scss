.wrapper {
  background: white;
  padding: 2rem 3rem;
  border-radius: 0.2rem;

  div {
    &:nth-child(2) {
      margin-top: 1rem;
    }
    &:nth-child(3) {
      margin-top: 3rem;
      text-align: center;
      color: #a4a4a4;
    }
  }
}
