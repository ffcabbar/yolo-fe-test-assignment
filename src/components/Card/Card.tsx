import { LazyQueryExecFunction, OperationVariables } from '@apollo/client';
import { useState } from 'react';
import { Button } from '../lib/Button/Button';
import { Input } from '../lib/Input/Input';
import styles from './Card.module.scss';

type Props = {
  getData: LazyQueryExecFunction<any, OperationVariables>;
};

export const Card = ({ getData }: Props) => {
  const [code, setCode] = useState('');

  const handleChange = (e: any) => {
    setCode(e.target.value);
  };

  const onClick = () => {
    if (code) {
      setCode('');
      getData({
        variables: { code }
      });
    } else {
      alert('enter a code please');
    }
  };

  return (
    <div className={styles.wrapper}>
      <div>
        <Input placeholder="BTC" value={code} onChange={handleChange} />
      </div>
      <div>
        <Button onClick={onClick}>Add</Button>
      </div>
      <div>
        <div>Use of this service is subject to terms and conditions.</div>
      </div>
    </div>
  );
};
