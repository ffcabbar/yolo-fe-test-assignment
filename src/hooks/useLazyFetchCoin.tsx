import { useLazyQuery } from '@apollo/client';
import { useEffect, useState } from 'react';
import { ICoin } from '../common/types';
import { GET_COIN_PRICE_QUERY } from '../graphql/Queries';
import { alertNotification } from '../helpers/alert-notification';
import { Notification } from '../common/enums';

export const useLazyFetchCoin = () => {
  const [coins, setCoins] = useState<ICoin[]>([]);

  useEffect(() => {
    const coins = localStorage.getItem('coinsInfo');
    if (coins) {
      setCoins(JSON.parse(coins));
    }
  }, []);

  const [getData, { loading, error }] = useLazyQuery(GET_COIN_PRICE_QUERY, {
    fetchPolicy: 'network-only',
    notifyOnNetworkStatusChange: true,
    onCompleted: (data) => {
      const hasSameCoin = coins.some((f) => f.id === data.markets[0]?.id);
      if (data.markets.length && !hasSameCoin) {
        const allCoins = [...coins, data.markets[0]];
        setCoins(allCoins);
        localStorage.setItem('coinsInfo', JSON.stringify(allCoins));
      } else if (data.markets.length <= 0) {
        alertNotification('Coin not found !', Notification.ERROR);
      }
      if (hasSameCoin) {
        alertNotification('This coin already exists on your list !', Notification.WARNING);
      }
    }
  });

  return { coins, setCoins, getData, loading, error };
};
