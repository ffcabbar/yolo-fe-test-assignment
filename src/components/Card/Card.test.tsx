import { MockedProvider } from '@apollo/react-testing';
import { renderHook, act } from '@testing-library/react-hooks';
import { useLazyFetchCoin } from '../../hooks/useLazyFetchCoin';
import { Card } from './Card';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';

const getHookWrapper = (mocks?: any, code?: string) => {
  const wrapper = ({ children }: any) => (
    <MockedProvider mocks={mocks} addTypename={false}>
      {children}
    </MockedProvider>
  );
  const { result, waitForNextUpdate } = renderHook(() => useLazyFetchCoin(), {
    wrapper
  });

  const { coins } = result.current;
  expect(typeof coins).toBe('object');

  act(() => {
    result.current.getData();
  });

  return { result, waitForNextUpdate };
};

describe('Card component', () => {
  it('should renders', async () => {
    const { result, waitForNextUpdate } = getHookWrapper();
    await waitForNextUpdate();

    const { getData } = result.current;
    const { container } = render(<Card getData={getData} />);
    expect(container).toBeInTheDocument();
  });

  it('matches snapshot', async () => {
    const { result, waitForNextUpdate } = getHookWrapper();
    await waitForNextUpdate();

    const { getData } = result.current;
    const tree = renderer.create(<Card getData={getData} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
