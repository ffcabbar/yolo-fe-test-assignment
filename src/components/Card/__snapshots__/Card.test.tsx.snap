// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component matches snapshot 1`] = `
<div
  className="wrapper"
  data-testid="card"
>
  <div>
    <div>
      <input
        className="input"
        data-testid="search-box"
        onChange={[Function]}
        placeholder="BTC"
        role="searchbox"
        value=""
      />
    </div>
  </div>
  <div>
    <button
      className="button"
      data-testid="button"
      onClick={[Function]}
    >
      Add
    </button>
  </div>
  <div>
    <div>
      Use of this service is subject to terms and conditions.
    </div>
  </div>
</div>
`;
